ledger/applications.go

```diff
@@ -235,3 +237,66 @@ func (al *logicLedger) DelGlobal(key string) error {
 func (al *logicLedger) GetDelta(txn *transactions.Transaction) (evalDelta basics.EvalDelta, err error) {
 	return al.cow.BuildEvalDelta(al.aidx, txn)
 }
+
+func (al *logicLedger) balances() (apply.Balances, error) {
+	balances, ok := al.cow.(apply.Balances)
+	if !ok {
+		return nil, fmt.Errorf("cannot get a Balances object from %v", al)
+	}
+	return balances, nil
+}
+
+func (al *logicLedger) pay(from basics.Address, to basics.Address, amount uint64, spec transactions.SpecialAddresses) error {
+	pay := transactions.PaymentTxnFields{
+		Receiver: to,
+		Amount:   basics.MicroAlgos{Raw: amount},
+	}
+	header := transactions.Header{
+		Sender: from,
+	}
+	balances, err := al.balances()
+	if err != nil {
+		return err
+	}
+	var ad transactions.ApplyData
+	err = apply.Payment(pay, header, balances, spec, &ad)
+	if err != nil {
+		return err
+	}
+	var empty transactions.ApplyData
+	if !ad.Equal(empty) {
+		return fmt.Errorf("Pay caused ad change %v", ad)
+	}
+	return nil
+}
+
+func (l *logicLedger) Perform(tx *transactions.Transaction, spec transactions.SpecialAddresses) error {
+	balances, err := l.balances()
+	if err != nil {
+		return err
+	}
+
+	var ad transactions.ApplyData
+	// move fee to pool
+	err = balances.Move(tx.Sender, spec.FeeSink, tx.Fee, &ad.SenderRewards, nil)
+	if err != nil {
+		return err
+	}
+
+	switch tx.Type {
+	case protocol.PaymentTx:
+		err = apply.Payment(tx.PaymentTxnFields, tx.Header, balances, spec, &ad)
+	case protocol.AssetTransferTx:
+		err = apply.AssetTransfer(tx.AssetTransferTxnFields, tx.Header, balances, spec, &ad)
+	default:
+		err = fmt.Errorf("%s tx in AVM", tx.Type)
+	}
+	if err != nil {
+		return err
+	}
+	if !ad.Equal(transactions.ApplyData{}) {
+		return fmt.Errorf("Perform caused ad change %v", ad)
+	}
+	return nil
```

Comment:
nice!
```golang
deltas := l.cow.deltas()
```
And then... repeat the same logic that is in accountUpdates? I wonder what is going to be written in ApplyData - new addr A balance / asset amount or a change? Full values are easier.

Please also do not forget to unit test the fact balance/asset changes from a cow created at `func (cb *roundCowState) StatefulEval` are propagated into parent cow and into a ledger eventually.
In this case ApplyData will continue serve informational purpose for indexer is it does now.

Reply:
This is just my stub code so that I can wrap my head around what can happen in the ad during one of these subtxn evaluations.  I hope nothing, but it will require more thought.  For example, if we allow closeTo, I guess it would write into ad.  If we send algos to a new address, I think it would update their rewards.

Reply:
This turned out a bit differently than I originally expected.  It became much easier to encode all the possible things that could happen in ApplyData by encoding the entire inner transactions, along with *their* ApplyData, rather than attempt to "flatten" the ADs into a top-level AD.